pennylane/qaoa/layers.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Methods that define cost and mixer layers for use in QAOA workflows.
+"""
+import pennylane as qml
+from pennylane.operation import Tensor
+
+
+def _diagonal_terms(hamiltonian):
+    r"""Checks if all terms in a Hamiltonian are products of diagonal Pauli gates (PauliZ and Identity).
+
+    Args:
+        hamiltonian (.Hamiltonian): The Hamiltonian being checked
+
+    Returns:
+        bool: ``True`` if all terms are products of diagonal Pauli gates, ``False`` otherwise
+    """
+    val = True
+
+    for i in hamiltonian.ops:
+        i = Tensor(i) if isinstance(i.name, str) else i
+        for j in i.obs:
+            if j.name != "PauliZ" and j.name != "Identity":
+                val = False
+
+    return val
+
+
+def cost_layer(hamiltonian):
+    r"""Returns the QAOA cost layer corresponding to a cost Hamiltonian.
+
+    For the cost Hamiltonian :math:`H_C`, this is defined as the following unitary:
+
+    .. math:: U_C \ = \ e^{-i \gamma H_C}
+
+    where :math:`\gamma` is a variational parameter.
+
+    Args:
+        hamiltonian (.Hamiltonian): The cost Hamiltonian
+
+    Raises:
+        ValueError: if the terms of the supplied cost Hamiltonian are not exclusively products of diagonal Pauli gates
+
+    .. UsageDetails::
+
+        We define a cost Hamiltonian and pass it into ``cost_layer``:
+        and pass it into ``cost_layer``:
+
+        .. code-block:: python
+
+            from pennylane import qaoa
+            import pennylane as qml
+
+            cost_h = qml.Hamiltonian([1, 1], [qml.PauliZ(0), qml.PauliZ(0) @ qml.PauliZ(1)])
+            cost_layer = qaoa.cost_layer(cost_h)
+
+        We can then use the cost layer within a quantum circuit:
+
+        .. code-block:: python
+
+            dev = qml.device('default.qubit', wires=2)
+
+            @qml.qnode(dev)
+            def circuit(gamma):
+
+                for i in range(2):
+                    qml.Hadamard(wires=i)
+
+                cost_layer(gamma)
+
+                return [qml.expval(qml.PauliZ(wires=i)) for i in range(2)]
+
+        which gives us a circuit of the form:
+
+        >>> circuit(0.5)
+
+        .. code-block:: none
+
+             0: ──H──RZ(-1.0)──╭RZ(-1.0)──┤ ⟨Z⟩
+             1: ──H────────────╰RZ(-1.0)──┤ ⟨Z⟩
```

Comment:
Nitpick: suggest making this all one (console) code block
```suggestion
        >>> circuit(0.5)
        >>> print(circuit.draw())
        0: ──H──RZ(-1.0)──╭RZ(-1.0)──┤ ⟨Z⟩
        1: ──H────────────╰RZ(-1.0)──┤ ⟨Z⟩
```